pyemma/coordinates/io/data_in_memory.py

```diff
@@ -130,37 +126,58 @@ def dimension(self):
         """
         return self.ndim
 
-    def reset(self):
+    def _reset(self):
         """Resets the data producer
         """
-        self.itraj = 0
-        self.t = 0
+        self._itraj = 0
+        self._t = 0
 
-    def next_chunk(self, lag=0):
+    def _next_chunk(self, lag=0):
         """
 
         :param lag:
         :return:
         """
-        # finished once with all trajectories? so reset the pointer to allow
+        # finished once with all trajectories? so _reset the pointer to allow
         # multi-pass
-        if self.itraj >= self.ntraj:
-            self.reset()
+        if self._itraj >= self.ntraj:
+            self._reset()
 
-        traj_len = self._lengths[self.itraj]
-        traj = self.data[self.itraj]
+        traj_len = self._lengths[self._itraj]
+        traj = self.data[self._itraj]
 
         # complete trajectory mode
         if self._chunksize == 0:
+            X = traj
+            self._itraj += 1
+
             if lag == 0:
-                X = traj
-                self.itraj += 1
                 return X
             else:
-                X = traj
                 Y = traj[lag:traj_len]
-                self.itraj += 1
                 return (X, Y)
+        # chunked mode
+        else:
+            upper_bound = min(self._t + self._chunksize, traj_len)
+            slice_x = slice(self._t, upper_bound)
+
+            X = traj[slice_x]
+            self._t += X.shape[0]
+
+            if lag == 0:
+                if self._t >= traj_len:
```

Comment:
move out of lag if statement and delete special case, since you always iterate over all data.